lib/wallaroo/core/common/state_step_creator.pony

```diff
@@ -0,0 +1,286 @@
+
+  fun ref add(state_name: String, key: Key, info: Info^) =>
+    try
+      _info.insert_if_absent(state_name, Map[Key, Info])?(key) = info
+    end
+
+  fun contains(state_name: String, key: Key): Bool =>
+    try
+      apply(state_name, key)?
+      true
+    else
+      false
+    end
+
+  fun clone(): KeyToStepInfoTag[Info] iso^ =>
+    let c = recover iso KeyToStepInfoTag[Info].create() end
+
+    for (sn, k_i) in _info.pairs() do
+      for (key, info) in k_i.pairs() do
+        c.add(sn, key, info)
+      end
+    end
+
+    c
+
+  fun pairs(): Iter[(String, String, Info)] =>
+    """
+    Return an iterator over tuples where the first two values are the state name
+    and the key, and the last value is the info value.
+    """
+    Iter[(String, Map[String, Info] box)](_info.pairs()).
+      flat_map[(String, (String, Info))](
+        { (k_m) => Iter[String].repeat_value(k_m._1)
+          .zip[(String, Info)](k_m._2.pairs()) }).
+      map[(String, String, Info)](
+        { (x) => (x._1, x._2._1, x._2._2) })
+
+actor StateStepCreator is Initializable
+  var _keys_to_steps: KeyToStepInfoTag[Step] = _keys_to_steps.create()
+  var _keys_to_step_ids: KeyToStepInfo[StepId] = _keys_to_step_ids.create()
+
+  let _step_id_gen: StepIdGenerator = _step_id_gen.create()
+
+  let _app_name: String
+  let _worker_name: String
+  let _metrics_conn: MetricsSink
+  let _auth: AmbientAuth
+  let _event_log: EventLog
+  var _recovery_replayer: (None | RecoveryReplayer) = None
+
+  var _outgoing_boundaries: Map[String, OutgoingBoundary] val =
+    recover _outgoing_boundaries.create() end
+
+  var _runner_builders: Map[String, RunnerBuilder] val =
+    recover _runner_builders.create() end
+
+  var _omni_router: (None | OmniRouter) = None
+
+  let _pending_steps: MapIs[Step, (String, Key, StepId)] =
+    _pending_steps.create()
+
+  var _router_registry: (None | RouterRegistry) = None
+
+  new create(auth: AmbientAuth,
+    app_name: String,
+    worker_name: String,
+    metrics_conn: MetricsSink,
+    event_log: EventLog)
+  =>
+    _auth = auth
+    _app_name = app_name
+    _worker_name = worker_name
+    _metrics_conn = metrics_conn
+    _event_log = event_log
+
+  be application_begin_reporting(initializer: LocalTopologyInitializer) =>
+    initializer.report_created(this)
+
+  be application_created(initializer: LocalTopologyInitializer,
+    omni_router: OmniRouter)
+  =>
+    _omni_router = omni_router
+
+  be application_initialized(initializer: LocalTopologyInitializer) =>
+    initializer.report_ready_to_work(this)
+
+  be application_ready_to_work(initializer: LocalTopologyInitializer) =>
+    None
+
+  be report_unknown_key(producer: Producer, state_name: String, key: Key) =>
+    if not _keys_to_steps.contains(state_name, key) then
+      @printf[I32]("State Step Creator should create step for key '%s'\n"
+        .cstring(), key.cstring())
+
+      try
+        (let recovery_replayer, let omni_router) = try
```

Comment:
If you don't use these more than once, can you handle this inline?

Reply:
This lets you constrain the try blocks more so that make debugging more straightforward.

Reply:
done